src/nvim/ui_compositor.c

```diff
@@ -101,6 +101,34 @@ bool ui_comp_should_draw(void)
   return composed_uis != 0 && valid_screen;
 }
 
+// Adjusts the position of a layer in the layers array based on its zindex.
+// Moves the layer either forward (if zindex increases) or backward (if zindex decreases).
+//
+// @param[in]  idx       The index of the current layer in the items array.
+// @param[in]  forward   True move forward otherwise move backward.
+void ui_comp_layers_adjust(size_t idx, bool forward)
+{
+  ScreenGrid **items = layers.items;
+  size_t size = layers.size;
+  ScreenGrid *layer = items[idx];
+
+  if (forward) {
+    while (idx < size - 1 && layer->zindex > items[idx + 1]->zindex) {
```

Comment:
nit: would it be clearer if we did not alias `layers.items`?
```suggestion
  ScreenGrid *layer = layers.items[idx];

  if (forward) {
    while (idx < size - 1 && layer->zindex > layers.items[idx + 1]->zindex) {
```